Marlin/src/gcode/calibrate/G28.cpp

```diff
@@ -411,11 +411,18 @@ void GcodeSuite::G28() {
     TERN_(HOME_Z_FIRST, if (doZ) homeaxis(Z_AXIS));
 
     const bool seenR = parser.seenval('R');
-    const float z_homing_height = seenR ? parser.value_linear_units() : Z_HOMING_HEIGHT;
+    float z_homing_height;
+    if (seenR)
+      z_homing_height = current_position.z + parser.value_linear_units();
+    else {
+      z_homing_height = (Z_HOMING_HEIGHT) - _MIN(probe.offset.z, 0);
+      if (!TERN(HOME_AFTER_DEACTIVATE, axis_is_trusted, axis_was_homed)(Z_AXIS))
+        z_homing_height += current_position.z;
+    }
```

Comment:
Note that `Z_HOMING_HEIGHT` used to be called `MIN_Z_HEIGHT_FOR_HOMING` and does not just apply to "Z homing," in case that is a point of any confusion. It actually applies to the minimum height that is safe for any XY movement during homing.

The `HOME_Z_FIRST` case above would typically be used when Z is homing to the top (or the bed moves down). In such a case, XY could then home right away, so on such a setup, `Z_HOMING_HEIGHT` might even be set to `Z_HOME_POS`.

At this point we don't know the true Z height unless Z is "trusted" so we want to move to (at minimum) the `Z_HOMING_HEIGHT`. If there is a setup where the bed falls down, we can trust Z is higher than expected. If the case was a `DELTA` where the nozzle could fall down, then we would want to always raise Z over the current position.

In practice, however, `DELTA` and other machines that have a Z carriage vulnerable to gravity are not allowed to sleep the steppers that keep Z in place.

Reply:
Yes, that's fine, I see the two heights are typically defaulted together:
https://github.com/MarlinFirmware/Marlin/blob/7e700c13948506ab0fc57719c7c1bd35f0cb33c4/Marlin/src/inc/Conditionals_post.h#L3182-L3196

The intention is to standardise the lift distance after probing, so that when a chain of commands like `G28;G29` are sent, so long as they're defined to the same height, the printer doesn't have to perform any vertical moves to tranisition between each them. So this should actually be `+ bltouch.z_extra_clearance()`, not `- _MIN(probe.offset.z, 0)` now I think about it.

But I see what you mean, the `_MIN()` is necessary, however I think this should actually be reworked to use `do_z_clearance_by()` instead now.

Reply:
Thinking about it further, it may make sense for it to be `Z_PROBE_SAFE_CLEARANCE + _MIN(probe.offset.z, 0)`, as with `probe_at_point()`.

Reply:
The former will result in a lift between G28 and G29 with HS mode enabled, but that seems correct.

Reply:
Implemented in cdc9973

Reply:
Using the `probe.offset.z` makes some sense, as a positive value would have the nozzle pressed against the bed when it triggers, so raising by that amount brings the nozzle to Z0, and raising more brings the nozzle the appropriate height above the bed.

Of course, if the user has their `probe.offset.z` correctly set, simply raising to the `Z…CLEARANCE` position already does the trick, but we have no guarantee of that. If the user has a messed up `probe.offset.z` all bets are off, but we should try to do the safe thing, so raising above the current position does always seem sensible. A weird `probe.offset.z` like +20 might cause an excessive raise, but at least it can't harm the bed.

We are so far only looking at the situation with the bed, of course, and not at situations where a probe might be used with a workpiece that is above 0. That kind of case will need more thought, as we have the choice to set Z to 0 in response to a probe, or just probe to get the height, so most of the time, again, it's preferable for the probe to raise above the current Z position.